app/script/view_model/ConversationInputEmojiViewModel.js

```diff
@@ -0,0 +1,190 @@
+      case z.util.KEYCODE.ESC:
+        this.emoji_list.remove();
+        this.emoji_start_pos = -1;
+        break;
+      case z.util.KEYCODE.ARROW_UP:
+      case z.util.KEYCODE.ARROW_DOWN:
+        this.rotate_emoji_list(event.keyCode === z.util.KEYCODE.ARROW_UP);
+        this.suppress_key_up = true;
+        break;
+      case z.util.KEYCODE.ENTER:
+        this.enter_emoji(event.target, this.emoji_list.find('.emoji.selected'));
+        break;
+      default:
+        return false;
+    }
+
+    event.preventDefault();
+    return true;
+  }
+
+  on_input_key_up(data, event) {
+    if (!this.suppress_key_up) {
+      const input = event.target;
+      const text = input.value || '';
+
+      if (text[input.selectionStart - 1] === ':') {
+        this.emoji_start_pos = input.selectionStart;
+        this.update_emoji_list(input);
+      } else if (this.emoji_start_pos !== -1) {
+        if (input.selectionStart < this.emoji_start_pos || text[this.emoji_start_pos - 1] !== ':') {
+          this.emoji_list.remove();
+          this.emoji_start_pos = -1;
+        } else {
+          this.update_emoji_list(input);
+        }
+      }
+    }
+
+    this.suppress_key_up = false;
+    return true;
+  }
+
+  update_emoji_list(input) {
+    if (!input.value) {
+      return;
+    }
+
+    const query = input.value.substr(this.emoji_start_pos, input.selectionStart - this.emoji_start_pos);
+    if (query.length < QUERY_MIN_LENGTH || query[0] === ' ' || !this.emoji_dict) {
+      this.emoji_list.remove();
+    } else {
+      const emoji_matched = this.emoji_dict
+        .filter((emoji) => emoji.indexOf(query) !== -1)
+        .slice(0, EMOJI_LIST_LENGTH)
+        .map((emoji) => {
+          const [code, name] = emoji.split('\t');
+          const parsed_unicode_emoji = String.fromCodePoint.apply(null, code.split(','));
+          return `<div class='emoji'><span class='symbol'>${parsed_unicode_emoji}</span>${name}</div>`;
+        })
+        .join('');
+
+      if (emoji_matched === '') {
+        this.emoji_list.remove();
+      } else {
+        this.emoji_list.html(emoji_matched).appendTo('body').show();
+        this.emoji_list.find('.emoji:nth(0)').addClass('selected');
+
+        const pos = this.get_cursor_pixel_pos(input);
+        const top = pos.top - this.emoji_list.height();
+
+        this.emoji_list.css('left', pos.left);
+        this.emoji_list.css('top', top);
+      }
+    }
+  }
+
+  rotate_emoji_list(backward) {
+    const previous = this.emoji_list.find('.emoji.selected');
+    const new_selection = (previous.index() + (backward ? -1 : 1)) % this.emoji_list.find('.emoji').length;
+    previous.removeClass('selected');
+    this.emoji_list.find(`.emoji:nth(${new_selection})`).addClass('selected');
+  }
+
+  enter_emoji(input, emoji_line) {
+    const emoji = emoji_line.find('.symbol').text();
+    const val = input.value;
+    input.value = val.substr(0, this.emoji_start_pos - 1) + emoji + val.substr(input.selectionStart);
```

Comment:
Here I would do the following:
1. Assign `val.substr(0, this.emoji_start_pos - 1)` to a variable with a meaningful name
1. Assign `val.substr(input.selectionStart);` to a variable with a meaningful name
1. Use a template string for the whole expression